heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManager.java

```diff
@@ -0,0 +1,385 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package com.twitter.heron.metricscachemgr;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+
+import com.twitter.heron.common.basics.Constants;
+import com.twitter.heron.common.basics.NIOLooper;
+import com.twitter.heron.common.basics.SysUtils;
+import com.twitter.heron.common.config.SystemConfig;
+import com.twitter.heron.common.network.HeronSocketOptions;
+import com.twitter.heron.common.utils.logging.ErrorReportLoggingHandler;
+import com.twitter.heron.common.utils.logging.LoggingHelper;
+import com.twitter.heron.metricscachemgr.metricscache.MetricsCache;
+import com.twitter.heron.metricsmgr.MetricsSinksConfig;
+import com.twitter.heron.proto.tmaster.TopologyMaster;
+import com.twitter.heron.spi.common.ClusterConfig;
+import com.twitter.heron.spi.common.Config;
+import com.twitter.heron.spi.common.Context;
+import com.twitter.heron.spi.common.Key;
+import com.twitter.heron.spi.statemgr.IStateManager;
+import com.twitter.heron.spi.utils.ReflectionUtils;
+
+/**
+ * main entry for metrics cache manager
+ * MetricsCacheManager holds three major objects:
+ * 1.MetricsCache metricsCache: it is the store where metrics and exceptions are cached.
+ * 2.MetricsCacheManagerServer metricsCacheManagerServer: it accepts the metric message from sink.
+ * 3.MetricsCacheManagerHttpServer metricsCacheManagerHttpServer: it responds to queries.
+ */
+public class MetricsCacheManager {
+  private static final Logger LOG = Logger.getLogger(MetricsCacheManager.class.getName());
+
+  private static final String METRICS_CACHE_HOST = "0.0.0.0";
```

Comment:
MetricsCacheManagerServer(in container 0) receives metric/exception message from sinks(in containers other than 0). They may not be in the same machine.